### PR TITLE
HDDS-6095. Skip slow integration tests, allow repeating known flaky tests 5x in CI

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

1. Create two special categories of integration tests using JUnit5 tags:

 * `@Flaky`: these are run in CI ~but failures are informational only~, and re-run max. 5 times in case of failure
 * `@Slow`: these are no longer run as part of CI

Tests can be tagged at class or method level.

Tagged tests still can be run locally as they are not `@Disabled`, nor put on `@Ignore`.  We can even create a scheduled job for running them periodically (e.g. once per day).  They are skipped in regular CI using Maven profiles.

Caveat: only JUnit5 tests are eligible.  JUnit4 tests need to be converted if we want to tag, which this change does for some tests.  `Test*Ozone*Filesystem*` needs [more work](https://issues.apache.org/jira/browse/HDDS-6152) before we can convert it to JUnit5.

2. Change most existing `@Ignore` annotations to one of these two new tags.

https://issues.apache.org/jira/browse/HDDS-6095

## How was this patch tested?

CI:
https://github.com/adoroszlai/hadoop-ozone/runs/5373197323#step:5:7

### CODE REVIEW

**Comment by adoroszlai** [comment]
@ghuangups please review

File: hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Annotation to mark test classes or methods with some intermittent failures.
+ * These are handled separately from the normal tests.  (Not required to pass,
+ * may be repeated automatically, etc.)
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Tag("flaky")
+public @interface Flaky {
```
Comment (ghuangups) [inline hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java:35]: Looks like the implementation is a little different from what we discussed. I thought we wanted to created 3 or 4 categories as: stable, probation, flakey and lengthy/slow.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestDirectoryDeletingServiceWithFSO.java [hunk 2]
```diff
@@ -140,6 +135,7 @@ public void cleanup() {
     }
   }
 
+  @Flaky("HDDS-6189")
```
Comment (ghuangups) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestDirectoryDeletingServiceWithFSO.java:138]: I'm not sure annotate at test code level is sustainable.

File: hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java:35]: Thanks @ghuangups for taking a look.

Initially I proposed to use text files listing tests for each category.  That's what we discussed offline.

In practice, however, I found it has a problem: using both include/exclude file and include/exclude list in POM (which we have for CI) is not really compatible.

Test list in text file also has two disadvantages compared to annotation-based tagging:
 * easier to make typo in test class/method name or pattern,
 * any renaming or other refactoring needs to be applied to the list as well, which is easy to forget.
Both of these are non-issues with tags.

Regarding the set of categories being defined:
 * Stable is everything that's not tagged, i.e. it is an implicit category.
 * Probation does not need to be tagged.  One can explicitly select any subset of the tests for execution via `-Dtest`, include file or Maven profile.  The CI-specific profiles are not active in this case.

Comment (ghuangups) [reply hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java:35]: Thanks @adoroszlai for the great explanation. Yes, sure, regarding the current categories implementation. 
Regarding the tagging, how about using a small tool to do the tagging at run time based on the text file for test listing? My concern is the sustainability. For example, any test fix or moving between categories or the like will involve code level changes and go through whole CI process. On the other hand, if we have a text file based list, that could be a lot light-weighted?

Comment (adoroszlai) [reply hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java:35]: > any test fix or moving between categories or the like will involve code level changes and go through whole CI process

I hope that any test fix involves code change. ;)

File: hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  /**
+   * The issue tracking the flaky test.
+   */
+  String[] value() default "";
```
Comment (errose28) [inline hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/tag/Flaky.java:39]: Do we want a default value for this? I think every flaky test should have a corresponding Jira so we don't ignore/forget about it.

**Comment by GeorgeJahad** [comment]
My understanding of this PR is that builds will no longer fail, if any flaky tests fail.  Is that correct?

My concern is that now we might make a change that will really break a "flaky" test, and we won't notice it.

What I would want, in a perfect world, is for any flaky test that fails the initial run to be run a few more times.  If they fail everytime, then something is really broken, and the ci should halt.

I can't see an easy way to do that in this PR, but I'm not that familiar with this code.  Do you think something like that would be possible?

**Comment by adoroszlai** [comment]
Thanks @GeorgeJahad for the comment.

> What I would want, in a perfect world, is for any flaky test that fails the initial run to be run a few more times. If they fail everytime, then something is really broken, and the ci should halt.

I plan to implement that in HDDS-6082.  Maybe I should just combine that change with this one.

**Comment by GeorgeJahad** [comment]
> I plan to implement that in [HDDS-6082](https://issues.apache.org/jira/browse/HDDS-6082).
That's awesome!  Thank you!

>  Maybe I should just combine that change with this one.
I don't think that is necessary as long as there isn't a long delay.

File: .github/workflows/post-commit.yml [hunk 4]
```diff
@@ -293,7 +293,10 @@ jobs:
             maven-repo-
       - name: Execute tests
         run: hadoop-ozone/dev-support/checks/integration.sh -P${{ matrix.profile }}
-        continue-on-error: ${{ matrix.profile == 'flaky' }}
+        if: matrix.profile != 'flaky'
+      - name: Execute flaky tests
+        run: hadoop-ozone/dev-support/checks/integration.sh -P${{ matrix.profile }} -Dsurefire.rerunFailingTestsCount=5 -Dsurefire.fork.timeout=3600
```
Comment (GeorgeJahad) [inline .github/workflows/post-commit.yml]: Very nice!  I didn't know about -Dsurefire.rerunFailingTestsCount

Comment (adoroszlai) [reply .github/workflows/post-commit.yml]: > So what happens if the test fails 5 times? does the whole run fail?

Yes, it fails in that case.  It's really 5 reruns, so 6 attempts total.

**Comment by adoroszlai** [comment]
Note: `TestHDDSUpgrade` never finishes in CI (even tried 5 hours timeout).  It fails much faster locally. ;)  So I had to mark the whole class as `@Slow` for now.

**Pull request merged**